### PR TITLE
Add document outlining safety mechanisms of java agent.

### PR DESCRIPTION
**Pull request by anuraaga**
I need a doc to link to for internal reviews about the safety of a java agent, a generally very scary concept. But ours has many nice safety mechanisms and this should be enough to assuage most fears.

### CODE REVIEW

**Comment by anuraaga** [comment]
And let me know if I missed anything nice.

**Comment by vovencij** [comment]
Nice doc! You definitely forgot new `matrix` and `smoke-test` combo, that actually attaches final agent jar to full-fledged app-servers running on different OS-es and JDKs.

**Comment by anuraaga** [comment]
@vovencij Indeed completely forgot to add smoke tests, thanks!

File: docs/safety-mechanisms.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# OpenTelemetry Java Agent Safety Mechanisms
+
+This document outlines the safety mechanisms we have in place to have confidence
+that the Java agent can be attached with a very low chance of affecting a user's
+app in a negative way, for example introducing crashes.
```
Comment (iNikem) [inline docs/safety-mechanisms.md:5]:
```suggestion
This document outlines the safety mechanisms we have in place to have confidence
that the Java agent can be attached to an user's application with a very low chance of 
affecting it in a negative way, for example introducing crashes.
```

**Review by iNikem — APPROVED** [review]
Thank you!

**Comment by trask** [comment]
👍👍

**Pull request merged**